agent/functional_tests/tests/functionaltests_unix_test.go

```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
 	err := networkModeTest(t, agent, "bridge")
 	require.NoError(t, err, "Networking mode bridge testing failed")
 }
+
+// TestCustomAttributesWithValidOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with a key, value pair
+func TestCustomAttributesWithValidOptions(t *testing.T) {
+	len := 25
+	key, value := GetRandomString(len), GetRandomString(len)
+
+	customAttribute := fmt.Sprintf("{\"%s\": \"%s\"}", key, value)
+	os.Setenv("ECS_INSTANCE_ATTRIBUTES", customAttribute)
+
+	agent := RunAgent(t, nil)
+	defer agent.Cleanup()
+
+	params := &ecsapi.DescribeContainerInstancesInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: []*string{&agent.ContainerInstanceArn},
+	}
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithEmptyValue tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with key and empty value
+func TestCustomAttributesWithEmptyValue(t *testing.T) {
+	len := 25
+	key, value := GetRandomString(len), ""
+
+	customAttribute := fmt.Sprintf("{\"%s\": \"%s\"}", key, value)
+	os.Setenv("ECS_INSTANCE_ATTRIBUTES", customAttribute)
+
+	agent := RunAgent(t, nil)
+	defer agent.Cleanup()
+
+	params := &ecsapi.DescribeContainerInstancesInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: []*string{&agent.ContainerInstanceArn},
+	}
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithMultipleOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with multiple key, value pairs
+func TestCustomAttributesWithMultipleOptions(t *testing.T) {
+	numAttributes := 2
+	customAttributes := `{
+		"key1": "val1",
+		"key0": "val0"
+	}`
+	os.Setenv("ECS_INSTANCE_ATTRIBUTES", customAttributes)
+
+	agent := RunAgent(t, nil)
+	defer agent.Cleanup()
+
+	params := &ecsapi.DescribeContainerInstancesInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: []*string{&agent.ContainerInstanceArn},
+	}
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	for i := 0; i < numAttributes; i++ {
+		k := "key" + strconv.Itoa(i)
+		v := "val" + strconv.Itoa(i)
+		assert.Equal(t, v, attribMap[k], "Values should match")
+	}
+}
+
+// TestCustomAttributesWithMaxOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with maximum number of supported key, value pairs
+func TestCustomAttributesWithMaxOptions(t *testing.T) {
```

Comment:
It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.